python/hsfs/core/vector_server.py

```diff
@@ -658,6 +697,204 @@ def filter_entry_by_join_index(self, entry, join_index):
+                        [
+                            (
+                                e.get(sk.required_serving_key)
+                                # Check if there is any entry matched with feature name,
+                                # if the required serving key is not provided.
+                                or e.get(sk.feature_name)
+                            )
+                            for sk in self._serving_key_by_serving_index[
+                                prepared_statement_index
+                            ]
+                        ]
+                    ),
+                    entries,
+                )
+            )
+            bind_params = {"batch_ids": entry_values_tuples}
+            entry_values.append(bind_params)
+
+        # run all the prepared statements in parallel using aiomysql engine
+        loop = asyncio.get_event_loop()
+        parallel_results = loop.run_until_complete(
+            self._launch_async_read(
+                list(prepared_statement_objects.values()), entry_values
+            )
+        )
+
+        # construct the results; same as before
+        for count, prepared_statement_index in enumerate(prepared_statement_objects):
+            # prepared_statement_index include fg with label only
+            # But _serving_key_by_serving_index include the index when the join_index is 0 (left side)
+            if prepared_statement_index not in self._serving_key_by_serving_index:
+                continue
+
+            statement_results = {}
+            serving_keys = self._serving_key_by_serving_index[prepared_statement_index]
+            # Use prefix from prepare statement because prefix from serving key is collision adjusted.
+            prefix_features = [
+                (self._prefix_by_serving_index[prepared_statement_index] or "")
+                + sk.feature_name
+                for sk in self._serving_key_by_serving_index[prepared_statement_index]
+            ]
+            # iterate over results by index of the prepared statement
+            for row in parallel_results[count]:
+                row_dict = dict(row)
+                # can primary key be complex feature?
+                result_dict = self.deserialize_complex_features(
+                    self._complex_features, row_dict
+                )
+                # note: should used serialized value
+                # as it is from users' input
+                statement_results[
+                    self._get_result_key(prefix_features, row_dict)
+                ] = result_dict
+
+            # add partial results to the global results
+            for i, entry in enumerate(entries):
+                batch_results[i].update(
+                    statement_results.get(
+                        self._get_result_key_serving_key(serving_keys, entry), {}
+                    )
+                )
+        return batch_results, serving_keys
+
+    async def _set_aiomysql_connection(self, options):
```

Comment:
Can you add an easy parameter for users to specify the level of parallelism (number of MySQL connections)?
The online_connector does not have the 'minsize' and 'maxsize' parameters for the connection pool size:

 pool = await aiomysql.create_pool(host=hostname, port=3306,
                                   user=uname, password=pwd, db=dbname, loop=loop, 
                                   minsize=4, maxsize=10)

Reply:
minsize and maxsize arguments can be set by users in dict `options` parameter. e.g. `init_serving(options={'minsize':1,'maxsize':1})` which is consistent with existing API